src/hotspot/share/prims/jvmtiExport.cpp

```diff
@@ -1596,7 +1594,33 @@ void JvmtiExport::post_method_exit(JavaThread *thread, Method* method, frame cur
         result = Handle(thread, oop_result);
       }
     }
+  }
+
+  if (exception_exit) {
+    post_method_exit_inner(thread, mh, state, exception_exit, current_frame, result, value);
```

Comment:
I think for exception exit, you also need JRT_BLOCK because you want the transition to thread_in_VM for this code, since JRT_BLOCK_ENTRY doesn't do the transition.  It should be safe for exception exit and retain the old behavior.

Reply:
Thanks for having a look coleen. In fact, not doing the JRT_BLOCK for the exception entry is intentional, because that entry goes through a different JRT_ENTRY (not JRT_BLOCK_ENTRY), that already transitions. So if I do the JRT_BLOCK for the exception path, it asserts saying hey you are already in VM.

Reply:
I tried that first, and ended up with a bunch of non-trivial code duplication instead, as reading the oop is done in both paths but for different reasons. One to preserve/restore it (interpreter remove_activation entry), but also inside of JvmtiExport::post_method_exit() so that it can be passed into the MethodExit. I will give it another shot and see if it is possible to refactor it in a better way.

Reply:
I uploaded a CR that does pretty much what you suggested, ish. Hope you like it!